File: uis/client/mobile_app/lib/services/account_service.dart
import 'dart:convert';
import 'package:flutter/foundation.dart' show kIsWeb;
import '../config/app_config.dart';
import 'api_service.dart';
import 'local_storage_service.dart';
import 'error_handler_service.dart';
import 'package:universal_html/html.dart' as html;

// Type definitions
class Account {
  final String id;
  final String userId;
  final String accountNumber;
  final String accountName;
  final String accountType; // 'savings' | 'current' | 'fixed_deposit' | 'domiciliary' | 'mint'
  final String currency;
  final double balance;
  final String status; // 'active' | 'inactive' | 'frozen'
  final bool isPrimary;
  final DateTime createdAt;

  Account({
    required this.id,
    required this.userId,
    required this.accountNumber,
    required this.accountName,
    required this.accountType,
    required this.currency,
    required this.balance,
    required this.status,
    required this.isPrimary,
    required this.createdAt,
  });

  factory Account.fromJson(Map<String, dynamic> json) {
    return Account(
      id: json['id'] as String,
      userId: json['user_id'] as String,
      accountNumber: json['account_number'] as String,
      accountName: json['account_name'] as String,
      accountType: json['account_type'] as String,
      currency: json['currency'] as String,
      balance: (json['balance'] ?? 0).toDouble(),
      status: json['status'] as String,
      isPrimary: json['is_primary'] ?? false,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }
}

class CreateAccountParams {
  final String userId;
  final String accountName;
  final String accountType;
  final String currency;

  CreateAccountParams({
    required this.userId,
    required this.accountName,
    required this.accountType,
    required this.currency,
  });
}

class AccountService {
  final ApiService _apiService = ApiService();

  /// Get all accounts for a user
  Future<List<Account>> getUserAccounts(String userId) async {
    try {
      final response = await _apiService.get('${AppConfig.accountEndpoint}?user_id=$userId');

      if (response.data['success'] == true) {
        final accountsData = response.data['data'] as List<dynamic>;
        return accountsData.map((acc) => Account.fromJson(acc as Map<String, dynamic>)).toList();
      }

      return [];
    } catch (error) {
      // ...existing code...
      rethrow;
    }
  }

  /// Create a new account
  Future<Map<String, dynamic>> createAccount(CreateAccountParams params) async {
    try {
      final response = await _apiService.post(
        '${AppConfig.accountEndpoint}/account',
        data: {
          'user_id': params.userId,
          'account_name': params.accountName,
          'account_type': params.accountType,
          'currency': params.currency,
        },
      );

      if (response.data['success'] == true) {
        return {
          'success': true,
          'message': response.data['message'] ?? 'Account created successfully',
          'data': response.data['data'] != null
              ? Account.fromJson(response.data['data'] as Map<String, dynamic>)
              : null,
        };
      }

      return {
        'success': false,
        'message': ErrorHandlerService.handleError(response.data),
      };
    } catch (error) {
      // ...existing code...
      return {
        'success': false,
        'message': ErrorHandlerService.handleError(error),
      };
    } catch (error) {
      // ...existing code...
      return {
        'success': false,
        'message': 'An error occurred. Please try again.',
      };
    }
  }

  /// Set an account as primary
  Future<Map<String, dynamic>> setPrimaryAccount(String accountId) async {
    try {
      final response = await _apiService.post('${AppConfig.accountEndpoint}/$accountId/set-primary');

      return {
        'success': response.data['success'] == true,
        'message': response.data['message'] ??
            (response.data['success'] == true ? 'Primary account updated' : 'Failed to set primary account'),
      };
    } catch (error) {
      // ...existing code...
      return {
        'success': false,
        'message': ErrorHandlerService.handleError(error),
      };
    } catch (error) {
      // ...existing code...
      return {
        'success': false,
        'message': 'An error occurred. Please try again.',
      };
    }
  }

  /// Freeze an account
  Future<Map<String, dynamic>> freezeAccount(String accountId) async {
    try {
      final response = await _apiService.post('${AppConfig.accountEndpoint}/$accountId/freeze');

      return {
        'success': response.data['success'] == true,
        'message': response.data['message'] ??
            (response.data['success'] == true ? 'Account frozen successfully' : 'Failed to freeze account'),
      };
    } catch (error) {
      // ...existing code...
      return {
        'success': false,
        'message': ErrorHandlerService.handleError(error),
      };
    } catch (error) {
      // ...existing code...
      return {
        'success': false,
        'message': 'An error occurred. Please try again.',
      };
    }
  }

  /// Unfreeze an account
  Future<Map<String, dynamic>> unfreezeAccount(String accountId) async {
    try {
      final response = await _apiService.post('${AppConfig.accountEndpoint}/$accountId/unfreeze');

      return {
        'success': response.data['success'] == true,
        'message': response.data['message'] ??
            (response.data['success'] == true ? 'Account unfrozen successfully' : 'Failed to unfreeze account'),
      };
    } catch (error) {
      // ...existing code...
      return {
        'success': false,
        'message': ErrorHandlerService.handleError(error),
      };
    } catch (error) {
      // ...existing code...
      return {
        'success': false,
        'message': 'An error occurred. Please try again.',
      };
    }
  }

  /// Get a single account by keycloak ID
  /// Matches web app: ${AppConfig.accountEndpoint}/account/keycloak/${keycloak_id}
  Future<Account?> getAccountByKeycloakId(String accountId) async {
    try {
      // Get keycloak_id from storage (matching web app: localStorage.getItem('keycloak_id'))
      String? keycloakId;
      if (kIsWeb) {
        try {
          keycloakId = html.window.localStorage['keycloak_id'];
        } catch (_) {
          // Skip if fails
        }
      } else {
        try {
          keycloakId = await LocalStorageService.getString('keycloak_id');
        } catch (_) {
          // Skip if fails
        }
      }
      
      if (keycloakId == null) {
        throw Exception('No keycloak_id found');
      }

      final response = await _apiService.get('${AppConfig.accountEndpoint}/keycloak/$keycloakId');
      print('Get Account Response: ${response.data}');

      if (response.data['account'] != null) {
        final acc = response.data['account'] as Map<String, dynamic>;
        
        // Store account data (matching web app: localStorage.setItem('account', acc))
        // Note: Web app stores account object directly, but we'll store as JSON string for mobile
        try {
          if (kIsWeb) {
            // Web app: localStorage.setItem('account', acc) - stores object directly
            // For Flutter web, we'll store as JSON string
            html.window.localStorage['account'] = jsonEncode(acc);
            if (acc['id'] != null) {
              html.window.localStorage['account_id'] = acc['id'].toString();
              // Also store in 'user' key for API service lookup (matching web app api_service.ts line 42)
              html.window.localStorage['user'] = jsonEncode({'account_id': acc['id']});
            }
          } else {
            // Store account as JSON string (matching web app behavior)
            await LocalStorageService.setString('account', jsonEncode(acc));
            if (acc['id'] != null) {
              await LocalStorageService.setString('account_id', acc['id'].toString());
              // Also store in 'user' key for API service lookup (matching web app api_service.ts line 42)
              await LocalStorageService.setString('user', jsonEncode({'account_id': acc['id']}));
            }
          }
        } catch (e) {
          print('Failed to store account data: $e');
        }
        
        return Account.fromJson(acc);
      }

      return null;
    } catch (error) {
      // ...existing code...
      rethrow;
    }
  }

  /// Setup transaction PIN
  /// Matches web app: POST /account/setup-pin
  Future<Map<String, dynamic>> setupPin(String pin) async {
    try {
      // Use /account/setup-pin (not accountEndpoint which is /account/account)
      final response = await _apiService.post(
        '/account/account/setup-pin',
        data: {
          'pin': pin,
        },
      );

      if (response.statusCode == 200 || response.statusCode == 201) {
        return {
          'success': true,
          'message': response.data['message'] ?? 'PIN setup successfully',
        };
      } else {
        return {
          'success': false,
          'message': ErrorHandlerService.handleError(response.data),
        };
      }
    } catch (e) {
      // ...existing code...
      return {
        'success': false,
        'message': ErrorHandlerService.handleError(e),
      };
    } catch (e) {
      // ...existing code...
      return {
        'success': false,
        'message': 'An error occurred. Please try again.',
      };
    }
  }

  /// Verify transaction PIN
  /// Matches web app: POST /account/verify-pin
  Future<Map<String, dynamic>> verifyPin(String pin) async {
    try {
      // Use /account/verify-pin (not accountEndpoint which is /account/account)
      final response = await _apiService.post(
        '/account/account/setup-pin',
        data: {
          'pin': pin,
        },
      );

      if (response.statusCode == 200) {
        return {
          'success': true,
          'message': response.data['message'] ?? 'PIN verified successfully',
        };
      } else {
        return {
          'success': false,
          'message': ErrorHandlerService.handleError(response.data),
        };
      }
    } catch (e) {
      // ...existing code...
      return {
        'success': false,
        'message': ErrorHandlerService.handleError(e),
      };
    } catch (e) {
      // ...existing code...
      return {
        'success': false,
        'message': 'An error occurred. Please try again.',
      };
    }
  }

  /// Change transaction PIN (verify current, then setup new)
  /// Matches web app: POST /account/verify-pin then POST /account/setup-pin
  Future<Map<String, dynamic>> changePin({
    required String currentPin,
    required String newPin,
  }) async {
    try {
      // First verify current PIN
      final verifyResult = await verifyPin(currentPin);
      if (!verifyResult['success']) {
        return {
          'success': false,
          'message': verifyResult['message'] ?? 'Current PIN is incorrect',
        };
      }

      // Then setup new PIN
      final setupResult = await setupPin(newPin);
      return setupResult;
    } catch (e) {
      // ...existing code...
      return {
        'success': false,
        'message': 'An error occurred. Please try again.',
      };
    }
  }
}

File: uis/client/web2/src/services/account_service.ts
import { AxiosError } from 'axios';
import { AppConfig } from '../config/app_config';
import { apiService } from './api_service';

export interface Account {
  id: string;
  userId: string;
  accountNumber: string;
  accountName: string;
  accountType: 'savings' | 'current' | 'fixed_deposit' | 'domiciliary' | 'mint';
  currency: string;
  balance: number;
  status: 'active' | 'inactive' | 'frozen';
  isPrimary: boolean;
  createdAt: Date;
}

export interface CreateAccountParams {
  userId: string;
  accountName: string;
  accountType: Account['accountType'];
  currency: string;
}

class AccountService {
  /**
   * Get all accounts for a user
   */
  async getUserAccounts(userId: string): Promise<Account[]> {
    try {
      const response = await apiService.get(`${AppConfig.accountEndpoint}?user_id=${userId}`);
      
      const data = response.data as { success?: boolean; data?: Record<string, unknown>[] };
      if (data.success) {
        const accountsData = data.data as Record<string, unknown>[];
        return accountsData.map((acc) => ({
          id: acc.id as string,
          userId: acc.user_id as string,
          accountNumber: acc.account_number as string,
          accountName: acc.account_name as string,
          accountType: acc.account_type as Account['accountType'],
          currency: acc.currency as string,
          balance: acc.balance as number,
          status: acc.status as Account['status'],
          isPrimary: acc.is_primary as boolean,
          createdAt: new Date(acc.created_at as string),
        }));
      }
      return [];
    } catch (error) {
      console.error('Failed to load accounts:', error);
      throw error;
    }
  }

  /**
   * Create a new account
   */
  async createAccount(params: CreateAccountParams): Promise<{ success: boolean; message: string; data?: Account }> {
    try {
      const response = await apiService.post(`${AppConfig.accountEndpoint}/account`, {
        user_id: params.userId,
        account_name: params.accountName,
        account_type: params.accountType,
        currency: params.currency,
      });

      const data = response.data as { success?: boolean; message?: string; data?: any };
      if (data.success) {
        return {
          success: true,
          message: data.message || 'Account created successfully',
          data: data.data ? {
            id: data.data.id,
            userId: data.data.user_id,
            accountNumber: data.data.account_number,
            accountName: data.data.account_name,
            accountType: data.data.account_type,
            currency: data.data.currency,
            balance: data.data.balance || 0,
            status: data.data.status || 'active',
            isPrimary: data.data.is_primary || false,
            createdAt: new Date(data.data.created_at),
          } : undefined,
        };
      }
      return {
        success: false,
        message: data.message || 'Failed to create account',
      };
    } catch (error: unknown) {
      console.error('Error creating account:', error);
      
      // Extract error message from axios error response
      let errorMessage = 'Error creating account';
      
      if (error instanceof AxiosError && error.response) {
        const errorData = error.response.data;
        if (errorData && typeof errorData === 'object') {
          if ('message' in errorData && typeof errorData.message === 'string') {
            errorMessage = errorData.message;
          } else if ('error' in errorData && typeof errorData.error === 'string') {
            errorMessage = errorData.error;
          }
        } else if (typeof errorData === 'string') {
          errorMessage = errorData;
        }
        
        // Check if it's a 401 error
        if (error.response.status === 401) {
          errorMessage = 'Unauthorized: Please ensure you are logged in and your session is valid. ' + 
                        (errorMessage !== 'Error creating account' ? errorMessage : '');
        }
      } else if (error instanceof Error) {
        errorMessage = error.message;
      }
      
      return {
        success: false,
        message: errorMessage,
      };
    }
  }

  /**
   * Set an account as primary
   */
  async setPrimaryAccount(accountId: string): Promise<{ success: boolean; message: string }> {
    try {
      const response = await apiService.post(`${AppConfig.accountEndpoint}/${accountId}/set-primary`);
      
      const data = response.data as { success?: boolean; message?: string };
      return {
        success: !!data.success,
        message: data.message || (data.success ? 'Primary account updated' : 'Failed to set primary account'),
      };
    } catch (error) {
      console.error('Error setting primary account:', error);
      return {
        success: false,
        message: error instanceof Error ? error.message : 'Error setting primary account',
      };
    }
  }

  /**
   * Freeze an account
   */
  async freezeAccount(accountId: string): Promise<{ success: boolean; message: string }> {
    try {
      const response = await apiService.post(`${AppConfig.accountEndpoint}/${accountId}/freeze`);
      
      const data = response.data as { success?: boolean; message?: string };
      return {
        success: !!data.success,
        message: data.message || (data.success ? 'Account frozen successfully' : 'Failed to freeze account'),
      };
    } catch (error) {
      console.error('Error freezing account:', error);
      return {
        success: false,
        message: error instanceof Error ? error.message : 'Error freezing account',
      };
    }
  }

  /**
   * Unfreeze an account
   */
  async unfreezeAccount(accountId: string): Promise<{ success: boolean; message: string }> {
    try {
      const response = await apiService.post(`${AppConfig.accountEndpoint}/${accountId}/unfreeze`);
      
      const data = response.data as { success?: boolean; message?: string };
      return {
        success: !!data.success,
        message: data.message || (data.success ? 'Account unfrozen successfully' : 'Failed to unfreeze account'),
      };
    } catch (error) {
      console.error('Error unfreezing account:', error);
      return {
        success: false,
        message: error instanceof Error ? error.message : 'Error unfreezing account',
      };
    }
  }

  /**
   * Get a single account by ID
   */
  async getAccountByKeycloakId(): Promise<Account | null> {
    try {
      const keycloak_id = localStorage.getItem('keycloak_id');
      const response = await apiService.get(`${AppConfig.accountEndpoint}/account/keycloak/${keycloak_id}`);
      
      const data = response.data as { success?: boolean; account?: any };
      if (data.success && data.account) {
        const acc = data.account;
        localStorage.setItem('account', JSON.stringify(acc));
        return {
          id: acc.id,
          userId: acc.user_id,
          accountNumber: acc.account_number,
          accountName: acc.account_name,
          accountType: acc.account_type,
          currency: acc.currency,
          balance: acc.balance,
          status: acc.status,
          isPrimary: acc.is_primary,
          createdAt: new Date(acc.created_at),
        };
      }
      return null;
    } catch (error) {
      console.error('Failed to load account:', error);
      throw error;
    }
  }
}

export const accountService = new AccountService();
